### DISCUSSION TITLE
Snapcast as "Output device" (sink) in remote pulseaudio

### DISCUSSION THREAD

**Discussion by sweenu** [snapcast/snapcast#992] [Q&A]
Hi !
I have my snapserver on a raspberry pi. I'd like to be able to stream whatever sound is playing on my computer to the snapserver.
I know with the tcp stream type, I can do something like `parec --rate=48000 --format=s16le --channels=2 | nc snapserver.home 4956` but from what I understand, it just lets me stream recording thourgh my input devices to snapcast.

What I'd really like is to be able to choose tcp streaming as an output (sink) like I would do for a speaker. Has anybody a clue on how to do that?

**Comment by kingosticks** [comment]
I am not sure I understand what you are asking here but did you look at streaming a PA "monitor" output using PA's native-protocol-tcp (with auth disabled)?

**Reply by sweenu** [reply]
So on the snapserver's config I added the stream `tcp://0.0.0.0:4953/?mode=client&name=PC` and did `pactl load-module module-native-protocol-tcp auth-anonymous=1`, also tried `pactl load-module module-simple-protocol-tcp  record=true port=4953 rate=48000 channels=2 format=s16le` but none of those did anything. I don't see any new stream in my snap frontend, and I don't see any new sink in pavucontrol on my computer.

**Reply by kingosticks** [reply]
I'm not sure what pulse's native/simple protocol entails but I guess they are more than just audio data, which is all snapserver tcp stream will ingest.

**Reply by kingosticks** [reply]
Actually, simple-protocol looks like what you want. You might need a `source` param specified.

`pactl load-module module-simple-protocol-tcp record=true port=4953 rate=48000 channels=2 format=s16le source=something`

Simple-protocol only sends PCM but snapcast default codec is flac so you'll need to change that.

`tcp://0.0.0.0:4953/?mode=client&codec=pcm&name=PC`

**Reply by sweenu** [reply]
Do you know how I can debug what's going on, on the client side? I load the module, but nothing seems to happen. I don't see any new log on the server side.
If I do `parec --rate=48000 --format=s16le --channels=2 -d=55 | nc 192.168.0.24 4953` I get: `write() failed: Broken pipe`.

**Reply by kingosticks** [reply]
>-d | --device=SINKORSOURCE
 >   Specify the symbolic name of the sink/source to play/record this stream on/from

55? Really?

**Reply by sweenu** [reply]
```
pactl list sources short
55      alsa_input.usb-046d_HD_Pro_Webcam_C920_21EE78FF-02.analog-stereo        PipeWire        s16le 2ch 32000Hz       SUSPENDED
56      alsa_output.usb-M-Audio_M-Track_Plus-00.analog-stereo.monitor   PipeWire        s32le 2ch 48000Hz       IDLE
57      alsa_input.usb-M-Audio_M-Track_Plus-00.analog-stereo    PipeWire        s32le 2ch 48000Hz       SUSPENDED
58      alsa_output.pci-0000_00_1f.3.analog-stereo.monitor      PipeWire        s32le 2ch 48000Hz       SUSPENDED
59      alsa_input.pci-0000_00_1f.3.analog-stereo       PipeWire        s32le 2ch 48000Hz       SUSPENDED
```

**Reply by sweenu** [reply]
It does not work currently, but even if it does work, what I understand is that it'll stream the received input from the microphone to snapcast. However, what I really want to do is stream whatever sound is playing on my computer (like a youtube video).

**Reply by kingosticks** [reply]
55 is not a symboolic **name**. It is the index. I'd guess the symbolic name is the delightfuly concise "alsa_input.usb-046d_HD_Pro_Webcam_C920_21EE78FF-02.analog-stereo". To use the index instead you probablty want:

> --monitor-stream=INDEX
 >   Record from the sink input with index INDEX.

And if you want to stream everything from your computer, you would use one of the "monitor" sources rather than a line-in/mic, just like I said at the start.

**Comment by sweenu** [comment]
I could have pulseaudio on the rpi and send my stream to the PA directly and then use that from snapcast but I'd very much prefer not using pulseaudio on my rpi and directly using the `tcp` stream.

**Comment by djmaze** [comment]
Trying to achieve the same. On my computer (here called `pc`), I ran this:

```bash
pactl load-module module-simple-protocol-tcp rate=48000 format=s16le channels=2 source=alsa_output.pci-0000_c5_00.6.analog-stereo record=true port=1714 listen=0.0.0.0
```

Using `nc pc 1714` on my Pi, I successfully get (audio) data back.

snapserver on my Pi is configured like this:

```
[stream]
source = tcp://pc:1714?name=pc&mode=client&codec=pcm
```

When starting snapserver on the Pi, I get the following output:

```
2025-07-14 15-12-41.058 [Debug] (parse) StreamUri: tcp://pc:1714?name=pc&mode=client&codec=pcm
2025-07-14 15-12-41.058 [Trace] (parse) scheme: 'tcp', tmp: '//pc:1714?name=pc&mode=client&codec=pcm'
2025-07-14 15-12-41.058 [Trace] (parse) host: 'pc:1714', tmp: '?name=pc&mode=client&codec=pcm', path: ''
2025-07-14 15-12-41.059 [Trace] (parse) path: '', tmp: 'name=pc&mode=client&codec=pcm', query: 'name=pc&mode=client&codec=pcm'
2025-07-14 15-12-41.059 [Debug] (parse) StreamUri.toString: tcp://pc:1714/?codec=pcm&mode=client&name=pc
2025-07-14 15-12-41.059 [Debug] (PcmStream) Start: pc, type: tcp, sampleformat: 48000:16:2, codec: pcm
2025-07-14 15-12-41.059 [Info] (PcmEnc) Init
2025-07-14 15-12-41.059 [Notice] (Server) Server::start: Invalid argument
2025-07-14 15-12-41.060 [Error] (Snapserver) Exception: Invalid argument
2025-07-14 15-12-41.061 [Notice] (Snapserver) Snapserver terminated.
```

Not sure what to make out of this. Is there a configuration problem? Is the audio data transmitted via Pulseaudio in the wrong format?

**Comment by fuchsmich** [comment]
Maybe airplay is an option for you?

Create an airplay sink in pipewire on your computer EDIT: or enable the RAOP Discover module in pipewire. 

Create an airplay source on your snapserver.

**Comment by sweenu** [comment]
That's a good idea, thank you!

**Reply by vjdw** [reply]
In case it helps, here are my notes from getting airplay working with snapcast on ubuntu

```sh
sudo apt install shairport-sync

# This leaves shairport-sync available to snapcast, but disables the standalone shairport service that might interfere.
sudo systemctl disable shairport-sync

```

Then add this to your `snapserver.conf` and restart snapserver

```config
source = airplay:///shairport-sync?name=AirPlay&devicename=MyAirPlay
```

At first it can seem like it's not working because the audio can take a few seconds to come through from the iPhone. If it still doesn't come through try increasing the volume on the iphone and waiting a couple of seconds. If it still doesn't work then maybe I missed something out of the configuration!
